lib/streamlit/runtime/state/query_params.py

```diff
@@ -0,0 +1,124 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass, field
+from typing import Any, Dict, Iterator, List, MutableMapping, Union
+
+from streamlit.proto.ForwardMsg_pb2 import ForwardMsg
+
+
+@dataclass
+class QueryParams(MutableMapping[str, Any]):
+    """A lightweight wrapper of a dict that sends forwardMsgs when state changes.
+    It stores str keys with str and List[str] values.
+    """
+
+    _query_params: Dict[str, Union[List[str], str]] = field(default_factory=dict)
+
+    def __init__(self):
+        # avoid using ._query_params as that will use __setattr__,
+        # which itself relies on `_query_params` being defined
+        self.__dict__["_query_params"] = {}
+
+    def __iter__(self) -> Iterator[Any]:
+        return iter(self._query_params.keys())
+
+    def __getitem__(self, key: str) -> str:
+        try:
+            value = self._query_params[key]
+            if isinstance(value, list):
+                if len(value) == 0:
+                    return ""
+                else:
+                    return value[-1]
+            return value
+        except:
+            raise KeyError(_missing_key_error_message(key))
+
+    def __setitem__(self, key: str, value: Union[str, List[Any]]) -> None:
+        # Handle non-string inputs like integers or lists by converting them to
+        # strings for consistent query parameter handling.
```

Comment:
Shouldn't the type of value be `Union[str, List[str]]`?

Also, what I think we want to get across in the comment is that the typecasting conflicts with the type signatures because we think being more strict for people that are using `mypy` is desirable. The current comment just sounds like a description of what we're doing that just duplicates the logic in the code